### PR TITLE
TorchScript BART models

### PR DESCRIPTION
**Pull request by stephenroller**
Enable TorchScripting of a BART-based TransformerGeneratorModel. Support for other Transformer variants may come at a later time.

### Known good conversation
Call to interact with model:
```
parlai interactive \
--model-file ${MODEL_FILE} \
--model bart \
--no-cuda
```
Output:
```
Enter Your Message: I am looking for a restaurant in the west part of town.
[Bart]: APICALL: Restaurant.area = west
Enter Your Message: APIRESP: Restaurant 14 matches
[Bart]: there are [value_count] restaurant -s in the [value_area] part of town . do you have a preference ?
```

### Exporting model

Call to export model:
```
parlai torchscript \
--model-file ${MODEL_FILE} \
--model bart \
--no-cuda \
--scripted-model-file ~/_test_scripted_model__bart.pt \
--input 'I am looking for a restaurant in the west part of town.|APIRESP: Restaurant 14 matches'
```
Output:
```
Generating given the original unscripted module:
 TEXT: I am looking for a restaurant in the west part of town.
LABEL: APICALL: Restaurant.area = west
 TEXT: APIRESP: Restaurant 14 matches
LABEL: there are [value_count] restaurant -s in the [value_area] part of town . do you have a preference ?

Generating given the scripted module:
 TEXT: I am looking for a restaurant in the west part of town.
LABEL: APICALL: Restaurant.area = west
 TEXT: APIRESP: Restaurant 14 matches
LABEL: there are [value_count] restaurant -s in the [value_area] part of town . do you have a preference ?
```

### Testing loading in and using exported model
Call:
```
python parlai/torchscript/scripts/test_exported_model.py \
--scripted-model-file ~/_test_scripted_model__bart.pt \
--input 'I am looking for a restaurant in the west part of town.|APIRESP: Restaurant 14 matches'
```
Output:
```
Generating given the scripted module:
 TEXT: I am looking for a restaurant in the west part of town.
LABEL: APICALL: Restaurant.area = west
 TEXT: APIRESP: Restaurant 14 matches
LABEL: there are [value_count] restaurant -s in the [value_area] part of town . do you have a preference ?
```

## Additional testing
- Unit test: `pytest tests/test_torchscript.py`
- Agent: `parlai interactive --model-file ~/_test_scripted_model__bart.pt --model parlai.torchscript.agents:TorchScriptAgent`

### CODE REVIEW

**Comment by EricMichaelSmith** [comment]
Ha, reminds me of the shenanigans I had to go through with ONNX (i.e. removing `FusedLayerNorm`, etc.)

**Comment by stephenroller** [comment]
> Ha, reminds me of the shenanigans I had to go through with ONNX (i.e. removing `FusedLayerNorm`, etc.)

ya that's what we need to discuss lol

File: projects/torchscript/util.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python3
```
Comment (stephenroller) [inline projects/torchscript/util.py:1]: let's move this whole thing into `parlai/torchscript` i guess, maybe `parlai/nn/torchscript` or something

Comment (EricMichaelSmith) [reply projects/torchscript/util.py]: I don't have a preference - moved into `parlai/torchscript/` for now but happy to revisit

File: .circleci/config.yml [hunk 2]
```diff
@@ -91,38 +91,41 @@ commands:
             python setup.py develop
             python -c "import nltk; nltk.download('punkt')"
 
-  installtorchgpu16:
+  installtorchgpu17:
```
Comment (stephenroller) [inline .circleci/config.yml:94]: let's rip all of circleci into a new PR which we can land much faster

File: parlai/agents/transformer/modules.py [hunk 3]
```diff
@@ -428,9 +430,9 @@ def forward_embedding(
             positions = (mask.cumsum(dim=1, dtype=torch.int64) - 1).clamp_(min=0)
         tensor = self.embeddings(input)
         if self.embeddings_scale:
-            tensor = tensor * np.sqrt(self.dim)
+            tensor = tensor * math.sqrt(self.dim)
 
-        if positions.max().item() > self.n_positions:
+        if positions.max().item() > self.n_positions and not torch.jit.is_scripting():
```
Comment (stephenroller) [inline parlai/agents/transformer/modules.py:435]: Can we swap the is_scriptings so they happen first? I suspect that will also help with the modelparallel madness. 

They're technically a microoptimization here computed to doing this cpu() transfer

File: parlai/agents/transformer/modules.py [hunk 4]
```diff
@@ -506,7 +511,7 @@ def forward(  # type: ignore
         input: torch.LongTensor,
         positions: Optional[torch.LongTensor] = None,
         segments: Optional[torch.LongTensor] = None,
-    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.BoolTensor]]:
+    ):
```
Comment (stephenroller) [inline parlai/agents/transformer/modules.py:514]:
```suggestion
    ):
    # TODO(spoff): Split TransformerEncoderLayer with and without reduction, and make type annotations always consistent
```

File: parlai/agents/transformer/modules.py [hunk 5]
```diff
@@ -727,10 +732,10 @@ def forward_embedding(
         """
         tensor = self.embeddings(input)
         if self.embeddings_scale:
-            tensor = tensor * np.sqrt(self.dim)
+            tensor = tensor * math.sqrt(self.dim)
         if self.variant == 'xlm':
             tensor = self.norm_embeddings(tensor)
-        if positions.max().item() > self.n_positions:
+        if positions.max().item() > self.n_positions and not torch.jit.is_scripting():
```
Comment (stephenroller) [inline parlai/agents/transformer/modules.py:738]: here too

File: parlai/agents/transformer/modules.py [hunk 6]
```diff
@@ -709,7 +714,7 @@ def _default(val, default):
     def forward_embedding(
         self,
         input: torch.LongTensor,
-        positions: Optional[torch.LongTensor] = None,
```
Comment (stephenroller) [inline parlai/agents/transformer/modules.py:712]: erm, IDK about this.

File: parlai/agents/transformer/modules.py [hunk 7]
```diff
@@ -766,23 +771,43 @@ def forward_layers(
...
             tensor, new_incr_state = self._apply_model_parallel(
                 tensor, encoder_output, encoder_mask, incr_state
             )
         else:
+            new_incr_state_by_layer = []
             for idx, layer in enumerate(self.layers):
-                tensor, new_incr_state[idx] = layer(
+                if incr_state is not None:
+                    single_layer_incr_state = {
+                        key: val[..., idx] for key, val in incr_state.items()
```
Comment (stephenroller) [inline parlai/agents/transformer/modules.py:786]: what is the ... syntax? Can you leave a comment explaining this?

File: parlai/agents/transformer/modules.py [hunk 8]
```diff
@@ -956,41 +1024,43 @@ def forward(self, x, encoder_output, encoder_mask, incr_state=None):
...
+        mask = torch.tril(torch.ones(time, time, dtype=x.dtype, device=x.device))
         # broadcast across batch
         mask = mask.unsqueeze(0).expand(bsz, -1, -1)
         return mask
 
-    def reorder_incremental_state(
-        self, incremental_state: Dict[str, dict], inds: torch.Tensor
-    ) -> Dict[str, dict]:
+
+class NegInfMixin:
```
Comment (stephenroller) [inline parlai/agents/transformer/modules.py:1048]: Sorry, I don't understand why this is a mixin.

Comment (stephenroller) [reply parlai/agents/transformer/modules.py:1048]: It feels obtuse. What other options do we have?

File: parlai/agents/transformer/modules.py [hunk 9]
```diff
@@ -1039,19 +1109,21 @@ def reorder_encoder_states(self, encoder_states, indices):
...
-        are dicts containing the incremental state for that layer.
+        Here, incremental_state is a dict whose keys are attention types (self- or
+        enc/dec and prev key, prev value, or prev mask) and whose values are Tensors,
+        with matrices for each layer concatenated in the last dimension. The 0th
+        dimension is the batch size and thus gets reindexed.
         """
         return {
-            idx: layer.reorder_incremental_state(incremental_state[idx], inds)
-            for idx, layer in enumerate(self.decoder.layers)
+            key: torch.index_select(val, 0, inds.to(val.device)).contiguous()
```
Comment (stephenroller) [inline parlai/agents/transformer/modules.py:1125]: I personally felt like the previous method was much cleaner. Why can't we torchscript those things?

File: parlai/torchscript/export_model.py [hunk 10]
```diff
@@ -0,0 +1,731 @@
+#!/usr/bin/env python3
```
Comment (stephenroller) [inline parlai/torchscript/export_model.py:1]: Oh lordy

File: parlai/torchscript/export_model.py [hunk 11]
```diff
@@ -0,0 +1,731 @@
...
+def export_model(opt: Opt):
+
+    opt['no_cuda'] = True
+    # TorchScripting is CPU only
+    opt['model_parallel'] = False
+    # model_parallel is not currently supported when TorchScripting
+
+    agent = create_agent(opt, requireModelExists=True)
+
+    inputs = opt['input'].split('|')
```
Comment (stephenroller) [inline parlai/torchscript/export_model.py:30]: ????

File: tests/test_torchscript.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        # Params
+        tasks = ['taskmaster2', 'convai2']
+        compiled_pattern = regex.compile(Gpt2BpeHelper.PATTERN)
+
+        with testing_utils.tempdir() as tmpdir:
+            datapath = tmpdir
+
+            for task in tasks:
+                parser = setup_args()
+                args = f"""\
```
Comment (stephenroller) [inline tests/test_torchscript.py:37]: I would prefer if we use the standard ParlaiScript methodology so we could do the more parlai-onic:

`JitExport.parse_kwargs(task=task, datatype='train:ordered', datapath=tmpdir)`

File: parlai/torchscript/export_model.py (hunk 11 above)
Comment (EricMichaelSmith) [reply parlai/torchscript/export_model.py:30]: This is just to provide sample input to test with when scripting

File: parlai/agents/transformer/modules.py (hunk 9 above)
Comment (EricMichaelSmith) [reply parlai/agents/transformer/modules.py:1125]: TorchScript wasn't able to handle the original Dict[Dict[Dict[str,Tensor]]] data structure

File: .circleci/config.yml (hunk 2 above)
Comment (EricMichaelSmith) [reply .circleci/config.yml:94]: Yes - see https://github.com/facebookresearch/ParlAI/pull/3521

File: parlai/agents/transformer/modules.py (hunk 3 above)
Comment (EricMichaelSmith) [reply parlai/agents/transformer/modules.py:435]: Yes, changing

File: parlai/agents/transformer/modules.py (hunk 6 above)
Comment (EricMichaelSmith) [reply parlai/agents/transformer/modules.py:712]: Yes, was able to revert this

File: parlai/agents/transformer/modules.py (hunk 7 above)
Comment (EricMichaelSmith) [reply parlai/agents/transformer/modules.py:787]: Yes, adding comment

File: parlai/agents/transformer/modules.py (hunk 8 above)
Comment (EricMichaelSmith) [reply parlai/agents/transformer/modules.py]: Okay, was able to revert this - this might be something that wasn't supported when scripting but is when tracing, I suspect

File: tests/test_torchscript.py (hunk 12 above)
Comment (EricMichaelSmith) [reply tests/test_torchscript.py]: Yes, changed

**Review by EricMichaelSmith — APPROVED** [review]
Trivially stamping since I've taken ownership of this PR (i.e. someone else should approve this too)

**Review by EricMichaelSmith — APPROVED** [review]

File: parlai/agents/jit/README.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
+# Agent exported to TorchScript (JIT compilation)
+
+This agent will read in a ParlAI agent that has been exported to TorchScript with JIT compilation, for use in greedy-search inference on CPU. This allows inference to be run on models without using any ParlAI overhead, either for tokenization or for the forward passes through the model. Currently, only BART models are supported.
```
Comment (klshuster) [inline parlai/agents/jit/README.md:3]: are there plans to support all `transformer/generator`s? if not, should we name this `bart_jit`?

File: parlai/agents/transformer/modules.py [hunk 14]
```diff
@@ -428,9 +429,9 @@ def forward_embedding(
             positions = (mask.cumsum(dim=1, dtype=torch.int64) - 1).clamp_(min=0)
         tensor = self.embeddings(input)
         if self.embeddings_scale:
-            tensor = tensor * np.sqrt(self.dim)
+            tensor = tensor * math.sqrt(self.dim)
```
Comment (klshuster) [inline parlai/agents/transformer/modules.py:432]: why this change? just curious

File: tests/test_jit.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+import parlai.utils.testing as testing_utils
+from parlai.agents.repeat_label.repeat_label import RepeatLabelAgent
+from parlai.core.agents import create_agent
+from parlai.core.params import ParlaiParser
+from parlai.core.worlds import create_task
+from parlai.utils.bpe import Gpt2BpeHelper
+
+
+@testing_utils.skipUnlessGPU
```
Comment (stephenroller) [inline tests/test_jit.py:23]: erm, why would you insist on a gpu for this test? Seems like the opposite based on comments

File: tests/test_jit.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+        with testing_utils.tempdir() as tmpdir:
+            datapath = tmpdir
+
+            for task in tasks:
+                opt = JitExport.setup_args().parse_kwargs(
+                    task=task, datatype='train:ordered', datapath=datapath
+                )
+                agent = RepeatLabelAgent(opt)
+                teacher = create_task(opt, agent).get_task_agent()
```
Comment (stephenroller) [inline tests/test_jit.py:45]:
```suggestion
                # TODO(roller): make a proper create_teacher helper
                teacher = create_task(opt, agent).get_task_agent()
```

File: parlai/agents/transformer/modules.py [hunk 17]
```diff
@@ -727,10 +733,10 @@ def forward_embedding(
         """
         tensor = self.embeddings(input)
         if self.embeddings_scale:
-            tensor = tensor * np.sqrt(self.dim)
+            tensor = tensor * math.sqrt(self.dim)
         if self.variant == 'xlm':
             tensor = self.norm_embeddings(tensor)
-        if positions.max().item() > self.n_positions:
+        if not torch.jit.is_scripting() and positions.max().item() > self.n_positions:
```
Comment (klshuster) [inline parlai/agents/transformer/modules.py:739]: why is this not supported by jit?

File: tests/test_jit.py [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+                        break
+
+    def test_jit_agent(self):
+        """
+        Test exporting a model to TorchScript and then testing it on sample data.
+        """
+
+        from parlai.scripts.jit_export import JitExport
+
+        test_phrase = "Don't have a cow, man!"  # From test_bart.py
```
Comment (stephenroller) [inline tests/test_jit.py:69]: lol

File: parlai/agents/transformer/modules.py [hunk 19]
```diff
@@ -918,12 +974,21 @@ def forward(self, x, encoder_output, encoder_mask, incr_state=None):
...
         x, final_self_attn_incr_state = self.self_attention(
             query=x,
+            key=None,
+            value=None,
             mask=decoder_mask,
-            incr_state=incr_state.get('self_attn'),
+            incr_state=self_attn_incr_state,
             static_kv=False,
         )[:2]
+        # Problematic to set None as default with type Optional[torch.Tensor]
```
Comment (klshuster) [inline parlai/agents/transformer/modules.py:991]: what is this referring to?

File: parlai/scripts/jit_export.py [hunk 20]
```diff
@@ -0,0 +1,751 @@
...
+
+
+@torch.jit.script
+class ScriptableGpt2BpeHelper(object):
+    """
+    Version of parlai.utils.bpe.Gpt2BpeHelper that can be TorchScripted.
+    """
+
+    @classmethod
+    def findall(cls, text: str) -> List[str]:
```
Comment (klshuster) [inline parlai/scripts/jit_export.py:328]: where is this method coming from?

**Review by klshuster — APPROVED** [review]
cool stuff

my one overall comment is - do we want to be a bit more forthcoming in the naming of these scripts/modules, that the only thing currently supported is BART?

File: tests/test_jit.py (hunk 15 above)
Comment (EricMichaelSmith) [reply tests/test_jit.py:23]: Because we only load fairseq for GPU tests :/ The alternative is to load fairseq for *all* of the tests

File: tests/test_jit.py (hunk 16 above)
Comment (EricMichaelSmith) [reply tests/test_jit.py:45]: Yes, added this (not committing this change since I've changed the path)

File: tests/test_jit.py (hunk 18 above)
Comment (EricMichaelSmith) [reply tests/test_jit.py:69]: ha took this from the existing BART test

File: parlai/agents/jit/README.md (hunk 13 above)
Comment (EricMichaelSmith) [reply parlai/agents/jit/README.md:3]: yeah, I definitely would like to see us support more than just BART, especially once we see how well TorchScripting works for the BART models

File: parlai/scripts/jit_export.py (hunk 20 above)
Comment (EricMichaelSmith) [reply parlai/scripts/jit_export.py:328]: I had to re-write regex for splitting tokens because you can't TorchScript it -__-

File: parlai/agents/transformer/modules.py (hunk 14 above)
Comment (EricMichaelSmith) [reply parlai/agents/transformer/modules.py:432]: Changing this back - some stuff you can't do when TorchScript scripting, but you *can* do it when TorchScript tracing, which we now do and has looser requirements

File: parlai/agents/transformer/modules.py (hunk 17 above)
Comment (EricMichaelSmith) [reply parlai/agents/transformer/modules.py:739]: Yes, looks like this is something else that is supported by tracing, so reverting this

File: parlai/agents/transformer/modules.py (hunk 19 above)
Comment (EricMichaelSmith) [reply parlai/agents/transformer/modules.py]: Never mind, reverted

File: parlai/agents/jit/jit.py [hunk 21]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply parlai/agents/jit/jit.py]: Yes, moving

File: parlai/scripts/torchscript.py [hunk 22]
```diff
@@ -0,0 +1,751 @@
...
+        for token in tokens:
+            decoded_token = self.decoder[token]
+            token_chars = self.utf8_chars(decoded_token)
+            for char in token_chars:
+                if not torch.jit.is_scripting():
+                    # We iterate over "char", which is supposed to be a single
+                    # character, because the TorchScripted version of the code
+                    # correctly splits a string into single characters in
+                    # self.utf8_chars() but the non-TorchScripted version doesn't
+                    chars.extend([real_char for real_char in char])
```
Comment (kauterry) [inline parlai/scripts/torchscript.py:578]:
```suggestion
                    chars.extend(list(real_char for real_char in char]))
```

**Comment by EricMichaelSmith** [comment]
@stephenroller Yes, I've re-split the incremental-state reordering logic back into TransformerDecoderLayer and MultiHeadAttention

File: tests/test_jit.py (hunk 18 above)
Comment (klshuster) [reply tests/test_jit.py]: Alt Text

File: parlai/scripts/torchscript.py (hunk 22 above)
Comment (klshuster) [reply parlai/scripts/torchscript.py]: @kauterry what's the rationale behind this suggestion? curious if it's better time/space complexity or something

Comment (kauterry) [reply parlai/scripts/torchscript.py]: My bad, I actually meant to suggest `chars.extend(list(char))`.
It's a style-based suggestion as you can see here:
Screen Shot 2021-03-17 at 2 04 33 PM
I guess we can skip this since @EricMichaelSmith says TorchScript throws an error with `list()`.

Comment (stephenroller) [reply parlai/scripts/torchscript.py]: the real suggestion is `chars.extend(list(char))`

Comment (klshuster) [reply parlai/scripts/torchscript.py]: great, thanks for the clarification makes sense!

Comment (EricMichaelSmith) [reply parlai/scripts/torchscript.py]: Just checked - actually, `list(char)` is fine, it's just the generator comprehension `(real_char for real_char in char)` that had an issue. Changed to `list(char)`

File: tests/test_jit.py (hunk 15 above)
Comment (EricMichaelSmith) [reply tests/test_jit.py]: Just changed now to use Kurt's use skip-unles-fairseq decorator
